src/aleph/vm/controllers/qemu/instance.py

```diff
@@ -0,0 +1,348 @@
+import asyncio
+import json
+import logging
+import shutil
+import sys
+from asyncio import Task
+from asyncio.subprocess import Process
+from typing import Generic, Optional, TypeVar
+
+import psutil
+import qmp
+from aleph_message.models import ItemHash
+from aleph_message.models.execution.environment import MachineResources
+from aleph_message.models.execution.instance import RootfsVolume
+from aleph_message.models.execution.volume import PersistentVolume, VolumePersistence
+
+from aleph.vm.conf import settings
+from aleph.vm.controllers.firecracker.executable import AlephFirecrackerResources
+from aleph.vm.controllers.interface import AlephControllerInterface
+from aleph.vm.controllers.qemu.cloudinit import CloudInitMixin
+from aleph.vm.network.firewall import teardown_nftables_for_vm
+from aleph.vm.network.interfaces import TapInterface
+from aleph.vm.storage import get_rootfs_base_path
+from aleph.vm.utils import HostNotFoundError, ping, run_in_subprocess
+
+logger = logging.getLogger(__name__)
+
+
+class AlephQemuResources(AlephFirecrackerResources):
+    async def download_all(self):
+        volume = self.message_content.rootfs
+        parent_image_path = await get_rootfs_base_path(volume.parent.ref)
+        self.rootfs_path = await self.make_writable_volume(parent_image_path, volume)
+        return
+
+    async def make_writable_volume(self, parent_image_path, volume: PersistentVolume | RootfsVolume):
+        "Create a new qcow2 image file based on the passed one, that we give to the VM to write onto"
+        qemu_img_path = shutil.which("qemu-img")
+        volume_name = volume.name if isinstance(volume, PersistentVolume) else "rootfs"
+
+        # detect the image format
+        out_json = await run_in_subprocess([qemu_img_path, "info", str(parent_image_path), "--output=json"])
+        out = json.loads(out_json)
+        parent_format = out.get("format", "")
```

Comment:
What happens below if `format` is empty?

Reply:
The next qemu-img command would fail. I will make a bit more safeguard in the python, it will provide more helpful debug information.